### PR TITLE
feat(sdk)!: retry broadcast operations

### PR DESCRIPTION
**Pull request by lklimek**
## Issue being fixed or feature implemented

Modification operations, like create identity etc., don't support retries even if DAPI returns retryable error.

## What was done?

* implemented Retry mechanism on StateTransition broadcast_and_wait, wait_for_response, broadcast.
* moved wait_timeout from broadcast_and_wait to PutSettings 
* replace some duplicate code with calls to StateTransition methods

## Breaking Changes

* BroadcastStateTransition::broadcast: added `Option<PutSettings>`
* BroadcastStateTransition::broadcast_and_wait: 
  *  removed timeout_ms
  * added `Option<PutSettings>`

## Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added or updated relevant unit/integration/functional/e2e tests
- [x] I have added "!" to the title and described breaking changes in the corresponding section if my code contains any
- [x] I have made corresponding changes to the documentation if needed

**For repository code-owners and collaborators only**
- [ ] I have assigned this pull request to a milestone

## Release Notes

- **New Features**
	- Introduced a new mechanism for wrapping execution results, enhancing error handling and response management.
	- Added a retry mechanism for broadcast requests, improving error management during execution.
	- New timeout configuration in `PutSettings` for better handling of state transition response waits.
	- Enhanced `BroadcastStateTransition` with configurable retry settings and new methods for handling responses.
	- Added documentation for the `transfer_credits` and `withdraw` methods to clarify their behavior.
	- Implemented a macro for generating gRPC request trait implementations, streamlining the process.
	- Enhanced `StateTransitionProofResult` with additional capabilities for formatted output and conversion.
	- Updated error handling in the `DapiClient` to improve logging around banned addresses.
	- Added a new error variant `AlreadyExists` for more informative error reporting.

- **Bug Fixes**
	- Improved error handling for asynchronous operations and added timeout support for state transition results.

- **Documentation**
	- Updated deprecation notices for the `CanRetry` trait, encouraging best practices in error handling.
	- Clarified documentation for the `timeout` field in `RequestSettings`.
	- Updated error message formats in test cases for consistency.

### CODE REVIEW

**Comment by lklimek** [comment]
@coderabbitai review all

File: packages/rs-sdk/src/platform/transition/put_contract.rs [hunk 1]
```diff
@@ -99,29 +87,7 @@ impl<S: Signer> PutContract<S> for DataContract {
...
-                )))?;
-
-        let (_, result) = Drive::verify_state_transition_was_executed_with_proof(
-            &state_transition,
-            &block_info,
-            proof.grovedb_proof.as_slice(),
-            &context_provider.as_contract_lookup_fn(),
-            sdk.version(),
-        )?;
+        let result = state_transition.wait_for_response(sdk, None).await?;
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-sdk/src/platform/transition/put_contract.rs:90]: out of scope of this PR

File: packages/rs-dapi-client/src/executor.rs [hunk 2]
```diff
@@ -145,3 +157,56 @@ where
...
+                address: result.address.clone(),
+            }),
+            Err(e) => ExecutionResult::Err(ExecutionError {
+                inner: e.into(),
+                retries: result.retries,
+                address: Some(result.address.clone()),
+            }),
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-dapi-client/src/executor.rs:212]: This trait also supports type conversions in the fly using From<> trait. It  is useful when using `?` to return errors.

File: packages/rs-dapi-client/src/dapi_client.rs [hunk 3]
```diff
@@ -227,7 +227,11 @@ impl DapiRequestExecutor for DapiClient {
                                     .address_list
                                     .write()
                                     .expect("can't get address list for write");
-
```
Comment (lklimek) [inline packages/rs-dapi-client/src/dapi_client.rs:230]: We need to fix banning on retry.

The issue is that on dapi-client layer, StaleNode issue is not an error, so the banning logic is not executed.
It becomes an issue on Sdk layer, so we need to ban it there.

**Review by shumkov — CHANGES_REQUESTED** [review]

File: packages/rs-dpp/src/state_transition/proof_result.rs [hunk 4]
```diff
@@ -5,7 +5,7 @@ use crate::voting::votes::Vote;
 use platform_value::Identifier;
 use std::collections::BTreeMap;
 
-#[derive(Debug)]
+#[derive(Debug, strum::Display, derive_more::TryInto)]
```
[the remark this answers is not in the document]
Comment (lklimek) [reply packages/rs-dpp/src/state_transition/proof_result.rs:8]: My goal is to just get a name of the enum variant selected; derive_more::Display requires either a #display("...") statement for each variant, or all fields to impement Display which is not the case here.

And `strum` was already here, I didn't add it.

File: packages/rs-dapi-client/src/dapi_client.rs (hunk 3 above)
Comment (lklimek) [reply packages/rs-dapi-client/src/dapi_client.rs:230]: TBD on separate PR

**Review by shumkov — DISMISSED** [review]

**Review by shumkov — APPROVED** [review]

**Pull request merged**